protocols/plaintext/src/handshake.rs

```diff
@@ -134,6 +134,12 @@ where
         }
     };
 
+    // The `Framed` wrapper may have buffered additional data that
+    // was already received but is no longer part of the plaintext
+    // handshake. We need to capture that data before dropping
+    // the `Framed` wrapper via `Framed::into_inner()`.
+    let read_buffer = framed_socket.read_buffer().clone().freeze();
```

Comment:
We may be able to avoid cloning here in the future, if something like https://github.com/matthunz/futures-codec/pull/54 is available.